Allow Asynchronous Calls in setup()/teardown() Functions

meanjoe45 opened the pull request:
This pull request is in reference to issue #70. To start off, this is a breaking change. When the `defer` option is set to true for a benchmark, both the `setup()` and `teardown()` functions require their resolve functions be called (`suResolve()` and `tdResolve()` respectively). If the functions are not included with the benchmark, the resolve calls are generated allowing `setup()` and `teardown()` functions to be omitted.

I tried to maintain the coding style as you describe, but have no problem changing variable names and such if you feel something else works better.

Currently, the unit tests are updated and are all passing when run with Node.js v4.x. I'm not very versed in running with the browser, so I'd appreciate any assistance people are willing to provide to test out these changes more. This includes verifying that the accuracy of the benchmarks has not been compromised.

Please reference the unit tests for examples of use. Documentation will need to be updated, but wanted to get some validation these changes will be accepted before investing the time to document.

Thanks for all the work you've put into benchmark.js.

meanjoe45 committed:
Initial support for asynchronous setup/teardown

meanjoe45 committed:
Add support for 'defer' without specifying asynchronous setup and/or teardown

meanjoe45 committed:
Update unit tests to account for asynchronous deferred setup/teardown

meanjoe45 committed:
Add unit test for asynchronous deferred setup/teardown

meanjoe45 commented:
I have run the unit tests with Node.js v4.4.2 and v6.9.1 locally on my machine and everything passes. I'm not sure yet why the CI build failed. I'll try to investigate soon, but help is appreciated.

meanjoe45 committed:
Fix function compiling taking Istanbul into account

shamasis commented:
@jdalton can we get this merged? this really a required feature where there is no option except async setup / teardown

daprahamian commented:
Is there a reason this has not been merged yet?

gwynjudd commented:
@jdalton is there anything that can be done to merge this? This is a very useful patch for doing async setup/teardown. Do you want to consider changing it so the breaking nature of the change doesn't occur?

Commit:
Fix benchmark tests when running with minified library code

This fixes an issue I found when using benchmark and my library was minified. I don't really understand all of the implications, but what seems to have happened is that the line "deferred.suResolve()" errors as "deferred" was undefined in minified.

By inspection, I noted that similar code was using the "interpolate" function and replaced "#d" with the name of the deferred variable.

shamasis commented:
@jdalton ping :-|

meanjoe45 committed:
Merge pull request #1 from gwynjudd/minfix

Fix benchmark tests when running with minified library code

vflopes commented:
+1
This feature is really awesome and I really need (I think a lot of people too).
Waiting for release.

rajsite commented:
We are trying to use benchmark.js for benchmarking the JS interface to some logic implemented in wasm and are also stuck because our wasm instantiation needs to happen asynchronously. Having async setup would definitely be helpful!

jszwedko commented:
Just bumping this as I ran into it today given the last movement was ~ a year ago. Is there anything that needs to change for this PR? I'm happy to try to pitch in.

Feirell commented:
@jdalton just another ping, I would help out if there is still something to do to make this happen.

timoxley commented:
Bump. This makes accurate async testing difficult

Uzlopak commented:
I will adapt this change to my fork

Finesse commented:
Waiting for the PR to be merged...

![Awaiter](https://icdn.lenta.ru/images/2017/02/02/15/20170202155432542/pic_c68a6b76fc08e7a072094aa7342242b2.jpg)
